File: Leetcode Problems/852.py
from typing import List

class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        n = len(arr)
        start = 1
        end = n-2

        while(start <= end):
            mid = start + (end - start) // 2

            if(arr[mid-1] < arr[mid] and arr[mid+1] < arr[mid]):
                return mid

            if(arr[mid] > arr[mid-1]):
                start = mid + 1
            
            else:
                end = mid - 1

        return -1
    
def main():
    print("Peak Index in Mountain Array")
    arr = [0, 1, 0]
    print(Solution().peakIndexInMountainArray(arr))
    arr = [24, 69, 100, 99, 79, 78, 67, 36, 26, 19]
    print(Solution().peakIndexInMountainArray(arr))
    
main()

File: Leetcode Problems/852.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int peakIndexInMountainArray(vector<int>& arr) {
        int n = arr.size();
        
        // for(int i=1;i<n-1;i++) {                                        // 1st Approach
        //     if(arr[i-1] < arr[i] && arr[i+1] < arr[i]) {
        //         return i;
        //     }
        // }

        // int start = 0;
        // int end = n-1;

        // while(start < end) {                                            // 2nd Approach
        //     int mid = start + (end - start) / 2;

        //     if(arr[mid] < arr[mid+1]) {
        //         start = mid + 1;
        //     }
        //     else {
        //         end = mid;
        //     }
        // }
        // return start;

        int start = 1;
        int end = n-2;

        while(start <= end) {
            int mid = start + (end - start) / 2;

            if(arr[mid - 1] < arr[mid] && arr[mid+1] < arr[mid]) {
                return mid;
            }

            if(arr[mid] < arr[mid+1]) {
                start = mid + 1;
            }
            else {
                end = mid - 1;
            }
        }

        return -1;
    }
};

int main() {
    Solution s;
    vector<int> arr = {0, 2, 1, 0};
    cout << s.peakIndexInMountainArray(arr) << endl;
}